### PR TITLE
Refactor SdsApi to support dynamic certificate validation context.

### PR DESCRIPTION
**Pull request by JimmyCYJ**

### CODE REVIEW

File: source/common/secret/sds_api.h [hunk 1]
```diff
@@ -43,14 +42,10 @@ class SdsApi : public Init::Target,
...
-  const Ssl::TlsCertificateConfig* secret() const override {
-    return tls_certificate_secrets_.get();
-  }
-
-  Common::CallbackHandle* addUpdateCallback(std::function<void()> callback) override {
-    return update_callback_manager_.add(callback);
-  }
+protected:
+  // Updates local storage of dynamic secrets and invokes callbacks.
+  virtual void updateConfigHelper(const envoy::api::v2::auth::Secret&) {}
```
Comment (lizan) [inline source/common/secret/sds_api.h]: this should be PURE, no?

File: source/common/secret/sds_api.h [hunk 2]
```diff
@@ -66,13 +61,73 @@ class SdsApi : public Init::Target,
...
+                       std::string sds_config_name, std::function<void()> destructor_cb)
+      : SdsApi(local_info, dispatcher, random, stats, cluster_manager, init_manager, sds_config,
+               sds_config_name, destructor_cb) {}
+
+  // SecretProvider
+  const Ssl::TlsCertificateConfig* secret() const override {
+    return tls_certificate_secrets_.get();
+  }
+
+  Common::CallbackHandle* addUpdateCallback(std::function<void()> callback) override {
```
Comment (lizan) [inline source/common/secret/sds_api.h]: move this and update_callback_manager to parent class?

File: source/common/secret/secret_manager_impl.cc [hunk 3]
```diff
@@ -76,22 +76,59 @@ TlsCertificateConfigProviderSharedPtr SecretManagerImpl::findOrCreateTlsCertific
...
+
+void SecretManagerImpl::removeDynamicCertificateValidationContextProvider(
+    const std::string& map_key) {
+  ENVOY_LOG(debug, "Unregister certificate validation context provider. hash key: {}", map_key);
+
+  auto num_deleted = dynamic_certificate_validation_context_providers_.erase(map_key);
+  ASSERT(num_deleted == 1, "");
+}
+
+CertificateValidationContextConfigProviderSharedPtr
```
Comment (qiwzhang) [inline source/common/secret/secret_manager_impl.cc]: I don't like the fact that many of code are duplicated.  But I don't have good solution either.

File: source/common/secret/sds_api.cc [hunk 4]
```diff
@@ -13,11 +13,12 @@
...
-               Runtime::RandomGenerator& random, Stats::Store& stats,
-               Upstream::ClusterManager& cluster_manager, Init::Manager& init_manager,
-               const envoy::api::v2::core::ConfigSource& sds_config, std::string sds_config_name,
-               std::function<void()> destructor_cb)
+template <class SecretType>
+SdsApi<SecretType>::SdsApi(const LocalInfo::LocalInfo& local_info, Event::Dispatcher& dispatcher,
+                           Runtime::RandomGenerator& random, Stats::Store& stats,
+                           Upstream::ClusterManager& cluster_manager, Init::Manager& init_manager,
+                           const envoy::api::v2::core::ConfigSource& sds_config,
+                           std::string sds_config_name, std::function<void()> destructor_cb)
```
Comment (lizan) [inline source/common/secret/sds_api.cc]: `const std::string&`

**Comment by JimmyCYJ** [comment]
Created a PR https://github.com/envoyproxy/envoy/pull/4355, close this one.

**Pull request closed without merging**